dir.go

```diff
@@ -331,17 +331,38 @@ func (d *Directory) Unlink(name string) error {
 	d.lock.Lock()
 	defer d.lock.Unlock()
 
-	delete(d.entriesCache, name)
+	// We need to decouple the `FSNode` from the rest of the tree
+	// so any modifications won't propagate upwards, it's not enough
+	// with just removing it from the cache.
+	entry, ok := d.entriesCache[name]
+	if ok {
+		delete(d.entriesCache, name)
+		entry.RemoveParent()
+	}
 
 	return d.unixfsDir.RemoveChild(d.ctx, name)
 }
 
+// TODO: Evaluate taking `parent` out of the `inode` which
+// was meant to be set at initialization only.
+// TODO: This could actually transform the directory into a
+// new root, but it wouldn't work with a `File`.
+func (d *Directory) RemoveParent() {
+	d.inode.parent = nil
```

Comment:
This isn't going to be thread-safe.

Reply:
Yes, I'm not even sure if we should even be manipulating the internal `inode`.